# Тести для малих вибірок та пропорцій



**Завдання 1**. E-commerce компанія після редизайну сайту підозрює, що **середній час до покупки (time-to-purchase)** користувачів **збільшився**.
Історично середній час від першого заходу на сайт до покупки становив $\mu_0 = 123.7$ хвилин.

Команда аналітиків випадково обрала дані **7 покупців після редизайну**:
`128, 135, 121, 142, 126, 151, 123`

З допомогою наявних даних зʼясуйте, чи збільшився середній час покупки після редизайну.

Для цього виконайте наступні 6 кроків. Правильне виконання кожного з кроків оцінюється в 1 бал.

1. Запишіть нульову та альтернативну гіпотези і визначте тип тесту.
2. Обчисліть вибіркові статистики: $\bar x$, $s$, $n$.
3. Оберіть тип тесту та виконайте його (знайдіть критичне значення тесту, статистику тесту та р-значення) будь-яким пасуючим способом, щоб перевірити гіпотезу на рівні значущості $\alpha = 0.10$.
4. Прийміть рішення, чи відхиляєте ви гіпотезу $H_0$ використовуючи p-value.
5. Напишіть висновок: чи справді редизайн сайту подовжив час до покупки?
6. Чи зміниться ваше рішення при зміні рівня значущості на $\alpha = 0.05$.


**1. Гіпотези**

H₀: μ = 123,7 (історичний середній час)

H₁: μ > 123,7 

Тип тесту: правосторонній (one-tailed), оскільки альтернативна гіпотеза стверджує, що середній час збільшився (тобто знаходиться справа на графіку).

In [1]:
# Метод 1. Ручний розрахунок

import numpy as np
from scipy import stats

# Дані з умови
avg_time = np.array([128, 135, 121, 142, 126, 151, 123])
mu0 = 123.7                                                # історичне середнє 
alpha = 0.10

# Крок 1. Обчислимо вибіркову статистику
x_bar = np.mean(avg_time)
s = np.std(avg_time, ddof=1)                               # стандартне відхилення з n-1
n = len(avg_time)

# Крок 2. Обчислимо t-статистику вручну
t_stat = (x_bar - mu0) / (s / np.sqrt(n))

# Крок 3. Критичне значення для правостороннього тесту
df = n - 1
t_crit = stats.t.ppf(1-alpha, df)

# Крок 4. p-value
p_value = 1-stats.t.cdf(t_stat, df)

print(f"Вибіркове середнє: {x_bar:.2f}")
print(f"Вибіркове стандартне відхилення: {s:.2f}")
print(f"t-статистика: {t_stat:.3f}")
print(f"Критичне значення (df={df}): {t_crit:.3f}")
print(f"p-value: {p_value:.4f}")

# Крок 5. Висновок
if p_value < alpha:
    print(f"Відхиляємо H0: середній час до покупки після редизайну сайту більше за 123,7$ на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо. Редизайн сайту не подовжив час до покупки")

Вибіркове середнє: 132.29
Вибіркове стандартне відхилення: 10.98
t-статистика: 2.069
Критичне значення (df=6): 1.440
p-value: 0.0420
Відхиляємо H0: середній час до покупки після редизайну сайту більше за 123,7$ на рівні значущості 0.1


In [2]:
# Метод 2. Scipy тест

res = stats.ttest_1samp(avg_time, popmean=mu0, alternative="greater")
print(f"scipy: t = {res.statistic:.3f}, p-value = {res.pvalue:.5f}")

# висновок на основі р-значення
if res.pvalue < alpha:
    print(f"Відхиляємо H0 на користь H1: є підстави вважати, що середній час до покупки після редизайну більший за {mu0} на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

scipy: t = 2.069, p-value = 0.04202
Відхиляємо H0 на користь H1: є підстави вважати, що середній час до покупки після редизайну більший за 123.7 на рівні значущості 0.1


In [3]:
# при зміні рівня значущості на alpha = 0.05
alpha = 0.05

if res.pvalue < alpha:
    print(f"Відхиляємо H0 на користь H1: є підстави вважати, що середній час до покупки після редизайну більший за {mu0} на рівні значущості {alpha}")
else:
    print("Не відхиляємо H0: доказів недостатньо")

Відхиляємо H0 на користь H1: є підстави вважати, що середній час до покупки після редизайну більший за 123.7 на рівні значущості 0.05


**Висновок:** 
За результатами одновибіркового правостороннього t-тесту отримано `t = 2.069` та `p-value = 0.0420`.

Оскільки `p-value < 0.10`, нульову гіпотезу відхиляємо на рівні значущості 10%.
Отже, маємо статистично значущі підстави вважати, що після редизайну середній час до покупки **збільшився**.

При рівні значущості `α = 0.05` висновок **не змінюється**, оскільки `p-value = 0.0420 < 0.05`. Тобто і на 5% рівні значущості також є підстави
стверджувати, що редизайн сайту подовжив середній час до покупки.

**Завдання 2.**

До спеціальної рекламної кампанії **23%** дорослих упізнавали логотип компанії. Після завершення кампанії відділ маркетингу провів опитування: з **1200** випадково відібраних дорослих **311** упізнали логотип.

Перевірте на рівні значущості **3%** ($\alpha=0.03$), чи дають ці дані достатні підстави стверджувати, що **тепер більше ніж 23%** дорослих упізнають логотип компанії. Для розвʼязку використовуйте бібліотеку `statsmodels`.

Зробіть висновок, чи зросла впізнаваність логотипу.

Додатково, обчисліть довірчий інтервал на заданому рівні значущості і проінтерпретуйте текстом - як він додатково пояснює прийняте нами рішення.

In [4]:
from statsmodels.stats.proportion import proportions_ztest, proportion_confint

count = 311      # кількість "успіхів"
nobs = 1200      # обсяг вибірки
p0 = 0.23        # гіпотетична пропорція (H0)
alpha = 0.03

# z-тест для однієї пропорції
stat, pval = proportions_ztest(count, nobs, value=p0, alternative='larger', prop_var=p0)
confint = proportion_confint(count, nobs, alpha=0.03, method='normal')                            # 97% довірчий інтервал

print("🔹 Тест для однієї пропорції")
print(f"Вибіркова пропорція = {count/nobs:.4f}")
print(f"Z-статистика = {stat:.3f}")
print(f"p-value = {pval:.5f}")
print(f"97% довірчий інтервал: ({confint[0]:.4f}, {confint[1]:.4f})")

if pval < alpha:
    print("Відхиляємо H0: є статистично значущі підстави вважати, що частка зросла і є більшою за 23%.")
else:
    print("Не відхиляємо H0: статистично значущих підстав стверджувати, що частка більша за 23%, недостатньо.")

🔹 Тест для однієї пропорції
Вибіркова пропорція = 0.2592
Z-статистика = 2.401
p-value = 0.00818
97% довірчий інтервал: (0.2317, 0.2866)
Відхиляємо H0: є статистично значущі підстави вважати, що частка зросла і є більшою за 23%.


**Висновок:** для перевірки гіпотези про зростання частки впізнаваності логотипу було використано одновибірковий z-test для пропорції.

Перевіряли гіпотези:

- **H₀:** \( p = 0.23 \)
- **H₁:** \( p > 0.23 \)

За результатами тесту отримано:

- вибіркова частка: \( 311/1200 = 0.2592 \) (25.92%)
- **z-статистика = 2.401**
- **p-value = 0.0082**

Оскільки **p-value < 0.03**, нульову гіпотезу відхиляємо на рівні значущості 3%.

Отже, маємо статистично значущі підстави вважати, що після рекламної кампанії частка дорослих, які впізнають логотип компанії, **стала більшою за 23%**.

Додатково було побудовано **97% довірчий інтервал** для частки:
**(0.2317; 0.2866)**, тобто від **23.17% до 28.66%**.

Оскільки весь довірчий інтервал лежить **вище за 23%**, це додатково підтверджує висновок про зростання впізнаваності логотипу.

**Time spent — 1.5 hour**

**What I learned**

- ознайомилась із новими тестами для **малих вибірок** та **пропорцій у популяції**;
- для **малих вибірок** навчилась використовувати **t-test**, перевіряти гіпотезу покроково вручну та за допомогою `scipy`;
- для **пропорцій у популяції** розібралась, як застосовувати **z-test для пропорції** в `statsmodels` і як додатково підтверджувати висновок через **довірчий інтервал**;
- закріпила, що рішення щодо гіпотези найзручніше приймати через порівняння **p-value** та **alpha**.